arch/riscv/core/smp.c

```diff
@@ -36,5 +38,58 @@ void z_riscv_secondary_cpu_init(int cpu_num)
 #ifdef CONFIG_PMP_STACK_GUARD
 	z_riscv_configure_interrupt_stack_guard();
 #endif
+	irq_enable(RISCV_MACHINE_SOFT_IRQ);
 	riscv_cpu_init[cpu_num].fn(riscv_cpu_init[cpu_num].arg);
 }
+
+#ifdef CONFIG_SMP
+static uintptr_t *get_hart_msip(int hart_id)
+{
+#ifdef CONFIG_64BIT
+	return (uintptr_t *)(uint64_t)(RISCV_MSIP_BASE + (hart_id * 4));
+#else
+	return (uintptr_t *)(RISCV_MSIP_BASE + (hart_id * 4));
+#endif
+}
+
+void arch_sched_ipi(void)
+{
+	unsigned int key;
+	uint32_t i;
+	uint8_t id;
+
+	key = arch_irq_lock();
+
+	id = _current_cpu->id;
+	for (i = 0U; i < CONFIG_MP_NUM_CPUS; i++) {
+		if (i != id) {
+			volatile uint32_t *r = (uint32_t *)get_hart_msip(i);
+			*r = 1U;
```

Comment:
What a pleasingly simple IPI!  If only other architectures could learn from this.  The x86 APIC isn't awful, I guess, but let me introduce you to cAVS IDC...